### PR TITLE
Better error messages for device_api to device interface failures.

### PR DESCRIPTION
**Pull request by zvookin**
This will help users recognize and fix argument errors around device
API and the target it is paired with. In particular, requesting a
default GPU without specifying one in the target is a common
error. This is currently only wired up to report errors when used
through Halide::Buffer, but that is the common case. (Use of
error_site is both to help pin down where the error occurred and to give
a parameter that preserves the current behavior of returning nullptr
as it is likely some code depends on that.)

Add tests for errors. (Couldn't figure out an easy way to test missing
the runtime so that is not covered.)

Tab cleanup in unrelated file (implicit_args.cpp)

(This is an alternative to https://github.com/halide/Halide/pull/2392 .)

### CODE REVIEW

File: src/DeviceInterface.h [hunk 1]
```diff
@@ -9,18 +9,17 @@
...
+ * DeviceAPI. If error_site is non-null, e.g. the name of the routine
+ * calling get_device_interface_for_device_api, a user_error is
+ * reported if the requested device API is not enabled in or supported
+ * by the target, Halide has been compiled without this device API, or
+ * the device API is None or Host or a bad value. The error_site
+ * argument is printed in the error message. If error_site is null,
+ * this routine returns nullptr instead of calling user_error. */
+EXPORT const halide_device_interface_t *get_device_interface_for_device_api(DeviceAPI d,
+                                                                            const Target &t = get_jit_target_from_environment(),
+                                                                            const char *error_site = nullptr);
```
Comment (dsharletg) [inline src/DeviceInterface.h]: I think this should have a non-null default, and maybe the function itself should assert it's non-null so we don't give silent failures.

File: test/error/implicit_args.cpp [hunk 2]
```diff
@@ -4,7 +4,7 @@
 using namespace Halide;
 
 int main(int argc, char **argv) {
-   	Var x("x"), y("y"), z("z");
+        Var x("x"), y("y"), z("z");
```
Comment (dsharletg) [inline test/error/implicit_args.cpp]: Weird indentation still

File: src/DeviceInterface.h (hunk 1 above)
Comment (zvookin) [reply src/DeviceInterface.h]: The return null behavior is used in at least one test to skip testing the particular GPU API if it is not enabled.

**Comment by steven-johnson** [comment]
Is this ready to land?

**Pull request merged**